File: algorithms/1624.largest-substring-between-two-equal-characters/solution.rs
use std::cmp::max;
use std::collections::HashMap;

impl Solution {
    pub fn max_length_between_equal_characters(s: String) -> i32 {
        let mut longest = -1;
        let mut map: HashMap<char, i32> = HashMap::new();

        for (i, c) in s.chars().enumerate() {
            let entry = map.entry(c).or_insert(-1);
            if *entry == -1 {
                *entry = i as i32;
            } else {
                longest = max(longest, i as i32 - *entry - 1);
            }
        }

        longest
    }
}

File: algorithms/1624.largest-substring-between-two-equal-characters/solution.cpp
#include <map>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int maxLengthBetweenEqualCharacters(string s) {
        int longest = -1;
        unordered_map<char, int> map;

        for (int i = 0; i < s.length(); i++) {
            char c = s[i];
            if (map.count(c)) {
                longest = max(longest, i - map[c] - 1);
            } else {
                map[c] = i;
            }
        }

        return longest;
    }
};
